packages/core/src/workflow/context.ts

```diff
@@ -1,31 +1,77 @@
-import { type EventTypes, type WorkflowEvent } from "./events";
-import { type StepFunction, type Workflow } from "./workflow";
+import { assertExists } from "../utils";
+import {
+  type EventTypes,
+  StartEvent,
+  StopEvent,
+  WorkflowEvent,
+} from "./events";
+
+export type StepFunction<
+  Start = string,
+  T extends WorkflowEvent = WorkflowEvent,
+> = (context: Context<Start>, ev: T) => Promise<WorkflowEvent>;
+
+export type StepMap = Map<
+  StepFunction<any>,
+  { inputs: EventTypes[]; outputs: EventTypes[] | undefined }
+>;
+
+export type ReadonlyStepMap = ReadonlyMap<
+  StepFunction<any>,
+  { inputs: EventTypes[]; outputs: EventTypes[] | undefined }
+>;
+
+export type ContextParams<Start> = {
+  startEvent: StartEvent<Start>;
+  steps: StepMap;
+  timeout: number | null;
+  verbose: boolean;
+};
+
+export class Context<Start = string>
+  implements AsyncIterable<WorkflowEvent, void, void>, Promise<StopEvent>
+{
+  readonly #steps: ReadonlyStepMap;
+  // reverse map of #steps, helper for get the next step
+  readonly #eventMap: WeakMap<typeof WorkflowEvent, StepFunction<Start>>;
+
+  readonly #startEvent: StartEvent<Start>;
+  readonly #queue: WorkflowEvent[] = [];
 
-export class Context {
-  #workflow: Workflow;
-  #queues: Map<StepFunction, WorkflowEvent[]> = new Map();
   #eventBuffer: Map<EventTypes, WorkflowEvent[]> = new Map();
-  #globals: Map<string, any> = new Map();
-  #streamingQueue: WorkflowEvent[] = [];
-  running: boolean = true;
+
+  #running: boolean = true;
+  #timeout: number | null = null;
   #verbose: boolean = false;
 
-  constructor(params: { workflow: Workflow; verbose?: boolean }) {
-    this.#workflow = params.workflow;
-    this.#verbose = params.verbose ?? false;
+  #getStepFunction(event: WorkflowEvent): StepFunction<Start> | undefined {
+    return this.#eventMap.get(event.constructor as EventTypes);
   }
 
-  set(key: string, value: any): void {
```

Comment:
I removed the globals here. because I think it's not really used in the code. Im thinking wheter to support this by adding `extend` syntax

Reply:
I don't see globals removed? But in any case, the get/set api (which stores stuff into globals) is a pretty useful api for sharing data between steps